### PR TITLE
Changed keySet from Set to ISet in ==>>

### PR DESCRIPTION
**Pull request by joshcough**
This is a bit more in the FP spirit.

I updated all the tests, and they all pass.

### CODE REVIEW

File: core/src/main/scala/scalaz/Map.scala [hunk 1]
```diff
@@ -220,8 +220,8 @@ sealed abstract class ==>>[A, B] {
   def keys: List[A] =
     foldrWithKey(List.empty[A])((x, _, xs) => x :: xs)
 
-  def keySet =
-    Set(keys: _*)
+  def keySet(implicit o: Order[A]): ISet[A] =
+    ISet.fromList(keys)
```
Comment (xuwei-k) [inline core/src/main/scala/scalaz/Map.scala]: Maybe `foldrWithKey(ISet.empty[A])((x, _, xs) => xs insert x)` is more efficient

Comment (puffnfresh) [inline core/src/main/scala/scalaz/Map.scala]: @xuwei-k :+1:

**Comment by S11001001** [comment]
This can be implemented without the `Order` constraint.  You can use `ISet`'s `Tip` and `Bin` data constructors, and match on the same of `==>>` to produce a matching set structure.  That way, there is no need to compare keys or rebalance the tree as it is constructed.

**Comment by joshcough** [comment]
Thanks for the comments. I will make the changes on Saturday.

**Comment by joshcough** [comment]
@S11001001 done. Thanks for the suggestion.

File: core/src/main/scala/scalaz/Map.scala (hunk 1 above)
Comment (joshcough) [inline core/src/main/scala/scalaz/Map.scala]: @xuwei-k i went with @S11001001 suggestion, since  it eliminated the `Order` constraint.

**Comment by S11001001** [comment]
:+1:

**Pull request merged**